### ISSUE TITLE
Extending the CSV reader/writer to support various complex data types

### ISSUE THREAD

**Issue by sighingnow** [v6d-io/v6d#761]
Describe your problem
---------------------

CSV is a commonly-used data format for data source/sink and vineyard currently depends on [apache-arrow][1] to parse and serialize tables to CSV files. The flexibility of the CSV parser is long-standing trouble for users of vineyard, e.g., the current CSV parser cannot handle nested data types (e.g., lists, dicts) well.

Inspired by the CSV parser and writer in other open-source projects (e.g., [ClickHouse][2]), vineyard decided to fork the CSV parser and writer from apache-arrow to support nested data types, including lists, maps, and structs. It's ok for users to referer to other projects' implementations and integrate them into the vineyard.

Related issues

- #236
- #248

Additional context
------------------

This issue is part of our [Alibaba Summer of Code 2022 Program](https://github.com/v6d-io/v6d/issues/758).

- Difficulty: Normal
- Mentor: Siyuan Zhang (@siyuan0322)

[1]: https://github.com/apache/arrow
[2]: https://clickhouse.com/docs/en/interfaces/formats/#tabseparated

**Comment by bobhan1** [comment]
Hi, I'm interested in this issue and want to handle this task for ASOC2022. Could you please provide more detailed information on the task and how can I get started on it? I want to make sure that if this task requires us to implement a standalone CSV parser in vineyard from scratch to support more complex data types(list and dict) instead of relying on `arrow::csv::TableReader` in Apache-arrow.

**Comment by sighingnow** [comment]
> I want to make sure that if this task requires us to implement a standalone CSV parser in vineyard from scratch to support more complex data types(list and dict) instead of relying on `arrow::csv::TableReader` in Apache-arrow.

Yes. The task is fairly standalone. You could add a csv parser/writer under the `modules/io/io` module and then we can replace the usages of arrow's CSV parser/writer in [io adaptors](https://github.com/v6d-io/v6d/blob/main/modules/io/io/kafka_io_adaptor.h).

You could borrow/refer the design of `arrow::csv::Table{Reader,Writer}` from apache-arrow.

**Comment by sighingnow** [comment]
> Could you please provide more detailed information on the task and how can I get started on it?

To get started, I suggest you first take a look at the implementation of the CSV reader/parser in the apache-arrow. And take a look at the reference implementation in other open source project, e.g., clickhouse, to get a sense about what we want to archieve.

**Comment by Nageshbansal** [comment]
@sighingnow, can we implement these drivers in python?

**Comment by Nageshbansal** [comment]
Hey! @sighingnow , can we implement the driver in python or go?

**Comment by sighingnow** [comment]
Comments from #248:

> We can get some ideas of supporting list/nested types with tsv/csv from https://clickhouse.com/docs/en/interfaces/formats/#tabseparated

**Comment by sighingnow** [comment]
> Hey! @sighingnow , can we implement the driver in python or go?

Hi @Nageshbansal,

I guess the answer is "No", as we need to integrate the reader as the backend for current `local_io_adaptor.cc` for efficiency guarantee.

Extending the arrow's CSV parser in C++ is the first choice, I think.

**Comment by Nageshbansal** [comment]
okay, then i will try to implement it in c++.
